test/test_other.py

```diff
@@ -8086,6 +8086,30 @@ def test_binaryen_warn_mem(self):
     self.run_process([EMCC, test_file('hello_world.c'), '-sINITIAL_MEMORY=' + str(16 * 1024 * 1024), '--pre-js', 'pre.js', '-sALLOW_MEMORY_GROWTH', '-sWASM_ASYNC_COMPILATION=0', '-sIMPORTED_MEMORY'])
     self.assertContained('hello, world!', self.run_js('a.out.js'))
 
+  def test_initial_heap(self):
+    for args, expected_initial_heap in [
+        ([], 16 * 1024 * 1024),         # Default behavior: 16MB initial heap
+        (['-sINITIAL_MEMORY=40MB'], 0), # Backwards compatibility: no initial heap (we can't tell if it'll fit)
+        (['-sMAXIMUM_MEMORY=40MB'], 0), # Backwards compatibility: no initial heap (we can't tell if it'll fit)
+        (['-sINITIAL_HEAP=64KB'], 64 * 1024), # Explicitly set initial heap is passed
+        (['-sINITIAL_HEAP=128KB', '-sINITIAL_MEMORY=20MB', '-sMAXIMUM_MEMORY=40MB'], 128 * 1024),
+        (['-sINITIAL_HEAP=10MB', '-sINITIAL_MEMORY=10MB'], -1), # Not enough space for stack
+        (['-sINITIAL_HEAP=5MB', '-sMAXIMUM_MEMORY=5MB'], -1), # Not enough space for stack
+      ]:
+      cmd = [EMCC, test_file('hello_world.c'), "-v"] + args
+      print(' '.join(cmd))
+
+      if expected_initial_heap < 0:
+        out = self.expect_fail(cmd)
+        self.assertContained('wasm-ld: error:', out)
+        continue
+
+      out = self.run_process(cmd, stderr=PIPE)
+      if expected_initial_heap != 0:
+        self.assertContained('--initial-heap=' + str(expected_initial_heap), out.stderr)
+      else:
+        self.assertNotContained('--initial-heap=', out.stderr)
```

Comment:
If possible, I'd prefer not to test this by peeking inside like this but I guess there is no other easy way?

I suppose we could compare `__heap_base` with `__builtin_wasm_memory_size` within the code?  But that might not work for == 0 case here.

Reply:
It was an intentional decision do this sort of "white box" testing. There were two reasons:

1) It's easier. As you mention, it is somewhat non-obvious how to test the `0` case.
2) It's composable in a sense that it tests only the functionality we're adding (i. e. we're not adding the feature that lays out a memory in a particular way, we're adding a feature that makes it such that a particular argument is passed to the linker).

The upside is that it is more resilient against changes in the other parts of the system. The downside is that it assumes the process involves invoking the linker, but it seems unlikely that the linker will go away any time soon (interestingly, writing a test with `__heap_base` would also assume a very particular linker implementation).